File: Dynamic Programming/416. Partition Equal Subset Sum(Recursion).cpp
class Solution {
public:

    int dp[205][20005];

    bool subset(int n, int sum, vector<int>& nums)
    {
        if(!sum) return true;
        if(!n) return false;
        if(dp[n][sum] != -1) return dp[n][sum];

        bool flag = subset(n-1, sum, nums);
        if(nums[n-1]<=sum)
            flag = flag or subset(n-1, sum - nums[n-1], nums);
        return dp[n][sum] = flag;
    }

    bool canPartition(vector<int>& nums) 
    {
        int sum = 0;
        for(int v:nums) sum += v;
        if(sum%2) return false;
        memset(dp, -1, sizeof(dp));
        return subset(nums.size(), sum/2, nums);
    }
};

File: Dynamic Programming/416. Partition Equal Subset Sum(Recursion).py
class Solution(object):
    def canPartition(self, nums):
        allSum = sum(nums)
        if allSum % 2: 
            return False

        memo = {}
        n = len(nums)
        target = allSum / 2
        def solve(i, curSum):
            if curSum == target:
                return True
            if curSum > target or i == n:
                return False

            if (i, curSum) in memo:
                return memo[(i, curSum)]

            pic = solve(i + 1, curSum + nums[i])
            not_pic = solve(i + 1, curSum)
            memo[(i, curSum)] = pic or not_pic
            return memo[(i, curSum)]

        return solve(0, 0)